fs: utils: copyfile: reset permissions on macos reflinks

efiop opened the pull request:
Related https://github.com/iterative/dvc-objects/pull/238

skshetry commented on src/dvc_objects/fs/utils.py:
```diff
@@ -189,6 +191,9 @@ def copyfile(
     try:
         system.reflink(src, dest)
+        if sys.platform == "darwin":
+            # NOTE: reflink on macos also clones src permissions
```
So to be clear, what pyreflink is doing is correct, then, right? It is always preserving permissions. It is not doing any chmod operations on macOS because clonefile does that already, but pyreflink is copying permissions on Linux as ficlone does not do that.

skshetry replied:
So, to be a 1:1 replacement with pyreflink, it has to be:

```python
reflink(src, dst)
if sys.platform == "linux":
    st = os.stat(src)
    os.chown(dst, st.st_uid, st.st_gid)
```

efiop replied:
I'm not sure I would say that cloning permissions is correct, as it seems inconsistent, a bit wasteful and also kinda unfinished, since it only copies mode but not other stuff (like shutil.copystats). E.g. for us the fact that pyreflink does that on linux is a complete waste, as we need to reset permissions ourselves anyway.

So yeah, we are chmodding different perms and for different reasons. And pyreflink is just trying to be consistent between linux and macos (thouhg I'm not sure if macos clone file clones stats in addition to just mode).

skshetry commented:
Did you test on Linux btw?

skshetry commented:
@efiop, looks like we need to chmod for linux too. About 10 tests are failing for me when I run with:
```console
TMPDIR=$(realpath ../tmp); mkdir $TMPDIR
pytest -nlogical
```

Most of those are import related and have `isexec: True` set. Setting `chmod` in both `copyfile` and `link` fix the issue for me.

skshetry commented:
For me,

```
reflink("pyproject.toml", "pyproject.toml2")
```

creates a file with executable bit set. Looks like this is a problem that I introduced with that optimization.

efiop commented:
@skshetry Thanks for testing it! I'll send a fix ASAP.

efiop commented:
@skshetry what filesystem do you use? I have a suspicion that it depends on the filesystem really (we tend to mixup platform and filesystem a lot when we discuss this, but it is really up to the fs on how it wants to handle it).

skshetry commented:
I am not sure what the fix should be. The `os.open` has a default mode of 077. `umask` is usually `022`. So the file gets created with `755` perms. 

One way might be to set `mode=0o666` in `os.open`, the other would be to copymode. But I am not sure where our fix should be here.

skshetry commented:
> @skshetry what filesystem do you use?

I use `btrfs`.

skshetry commented:
I see `builtin.open()` sets `0o666` by default. I think I should do the same.

https://github.com/python/cpython/blob/cc7e45cc572dd818412a649970fdee579417701f/Lib/_pyio.py#L1550

efiop commented:
> Looks like this is a problem that I introduced with that optimization.

I'm pretty sure it was me, I removed chmod that we used to have completely and only kept it for macos after that.

efiop commented:
@skshetry We just use the umask, so the file we create is the same as what the user would create from scratch. Same as shutil.copyfile too. The normal way to go about it is to use umask anyway, so I don't think we need to do anything special here and pyreflink's approach actually looks wrong to me.

skshetry commented:
yeah. I'd still sleep better today if we merge this https://github.com/iterative/dvc-objects/pull/248.

skshetry commented:
yes, everything works, all the tests pass when I change tmpdir to btrfs.

efiop commented:
@skshetry Didn't realise it was setting that mode before cloning. Indeed, you are correct about that.
